pubtools/_impl/tracing.py

```diff
@@ -0,0 +1,139 @@
+"""Instrument functions.
+
+Usage:
+    @instrument_func()
+      def func():
+          pass
+
+"""
+import functools
+import logging
+import os
+import threading
+
+from opentelemetry import baggage, context, trace
+from opentelemetry.baggage.propagation import W3CBaggagePropagator
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import (BatchSpanProcessor,
+                                            ConsoleSpanExporter)
+from opentelemetry.trace import Status, StatusCode
+from opentelemetry.trace.propagation.tracecontext import \
+    TraceContextTextMapPropagator
+
+from pubtools.pluggy import pm
+
+propagator = TraceContextTextMapPropagator()
+baggage_propagator = W3CBaggagePropagator()
+log = logging.getLogger(__name__)
+
+
+def get_trace_wrapper():
+    """return a global trace wrapper instance"""
+    return TracingWrapper()
+
+
+class TracingWrapper:
+    """Wrapper class to initialize opentelemetry instrumentation and provide a helper function
+    for instrumenting a function"""
+
+    def __new__(cls):
+        if not hasattr(TracingWrapper, "instance"):
+            cls.instance = super().__new__(cls)
+            if os.getenv("OTEL_TRACING", "").lower() == "true":
+                log.info("Creating TracingWrapper instance")
+                exporter = pm.hook.otel_exporter() or ConsoleSpanExporter()
+                cls.provider = TracerProvider(
```

Comment:
Is there a particular reason that you set these on the class object rather than the instance just created?

I suppose that it will work, but it's odd and it's effectively creating two levels of singleton.  In the same way that TracingWrapper.instance is global, so is TracingWrapper.provider, TracingWrapper.processor, TracingWrapper.tracer.  It's questionable then why a class is even needed if there is never any instance-specific data.

This should all work but I think it would be way cleaner if it were written like the below.  Just implement TracingWrapper as a plain old class, no overriding `__new__`, and then implement get_trace_wrapper() so it returns an existing copy of it.

```
TRACE_WRAPPER = None

def get_trace_wrapper():
    """return a global trace wrapper instance"""
    if TRACE_WRAPPER is None:
        TRACE_WRAPPER = TracingWrapper()
    return TRACE_WRAPPER

# TracingWrapper is implemented "normally"
class TracingWrapper:
    def __init__(self):
        # These should be private. We want callers to only use our
        # public methods e.g. `force_flush` so we can handle the `tracing disabled` case
        # for them.
        self._processor = None
        self._provider = None
        self._tracer = None
        if os.getenv("OTEL_TRACING", "").lower() == "true":
            # OK, actually initialize self._processor etc to functioning objects now
            self._processor = ...

    def force_flush(self):
        """Flush trace data into OTEL collectors"""
        if self._processor:
            self._processor.force_flush()
```

Personally I think that's a lot easier to understand and deal with, and it means that tests in this repo can just write `TracingWrapper()` and have it work as expected - creating a clean new instance of the object.

Also, we haven't discussed it before, but the singleton initialization would need to use a lock if there is any chance that multiple threads might be doing the initialization at the same time.

Reply:
> Is there a particular reason that you set these on the class object rather than the instance just created?

No special reason. Just thought that TracingWrapper is only needed to instanced once in the pubd/pub task job and referred it to other projects, so use singleton. Implementing TracingWrapper as a plain old class looks better, let me update it.